src/Microsoft.ML.CpuMath/FactorizationMachine/AvxIntrinsics.cs

```diff
@@ -0,0 +1,188 @@
+
+namespace Microsoft.ML.Internal.CpuMath.FactorizationMachine
+{
+    internal static class AvxIntrinsics
+    {
+        // This function implements Algorithm 1 in https://github.com/wschin/fast-ffm/blob/master/fast-ffm.pdf.
+        // Compute the output value of the field-aware factorization, as the sum of the linear part and the latent part.
+        // The linear part is the inner product of linearWeights and featureValues.
+        // The latent part is the sum of all intra-field interactions in one field f, for all fields possible
+        public static unsafe void CalculateIntermediateVariables(int fieldCount, int latentDim, int count, int* fieldIndices, int* featureIndices, float* featureValues,
+            float* linearWeights, float* latentWeights, float* latentSum, float* response)
+        {
+            // The number of all possible fields.
+            int m = fieldCount;
+            int d = latentDim;
+            int c = count;
+            int* pf = fieldIndices;
+            int* pi = featureIndices;
+            float* px = featureValues;
+            float* pw = linearWeights;
+            float* pv = latentWeights;
+            float* pq = latentSum;
+            float linearResponse = 0;
+            float latentResponse = 0;
+
+            Unsafe.InitBlock(pq, 0, (uint)(m*m*d*sizeof(float)));
+
+            Vector256<float> y = Vector256<float>.Zero;
+            Vector256<float> tmp = Vector256<float>.Zero;
+
+            for (int i = 0; i < c; i++)
+            {
+                int f = pf[i];
+                int j = pi[i];
+                linearResponse += pw[j] * px[i];
+
+                Vector256<float> x = Avx.BroadcastScalarToVector256(px + i);
+                Vector256<float> xx = Avx.Multiply(x, x);
+
+                // tmp -= <v_j,f, v_j,f> * x * x
+                int vBias = j * m * d + f * d;
+
+                // j-th feature's latent vector in the f-th field hidden space.
+                float* vjf = pv + vBias;
+
+                for (int k = 0; k + 8 <= d; k += 8)
+                {
+                    Vector256<float> vjfBuffer = Avx.LoadVector256(vjf + k);
+                    tmp = Avx.Subtract(tmp, Avx.Multiply(Avx.Multiply(vjfBuffer, vjfBuffer), xx));
+                }
+
+                for (int fprime = 0; fprime < m; fprime++)
+                {
+                    vBias = j * m * d + fprime * d;
+                    int qBias = f * m * d + fprime * d;
+                    float* vjfprime = pv + vBias;
+                    float* qffprime = pq + qBias;
+
+                    // q_f,f' += v_j,f' * x
+                    for (int k = 0; k + 8 <= d; k += 8)
+                    {
+                        Vector256<float> vjfprimeBuffer = Avx.LoadVector256(vjfprime + k);
+                        Vector256<float> q = Avx.LoadVector256(qffprime + k);
+                        q = Avx.Add(q, Avx.Multiply(vjfprimeBuffer, x));
+                        Avx.Store(qffprime + k, q);
```

Comment:
What's the codegen for this sequence... I would hope it folds both loads, but I've seen some codegen here sometimes.

Reply:
Codegen: 

```
movsxd rbx, eax
vmovups ymm3, ymmword ptr [r8+rbx*4]
movsxd rbx, eax
vmovups ymm4, ymmword ptr [r10+rbx*4]
vmulps ymm3, ymm3, ymm2
vaddps ymm4, ymm4, ymm3
movsxd rbx, eax
vmovups ymmword ptr [r10+rbx*4], ymm4
```

Reply:
> What's the codegen for this sequence... I would hope it folds both loads, but I've seen some codegen here sometimes.

@tannergooding 
Should this be created as an issue in coreclr?

Reply:
I already logged one for it a while back: <https://github.com/dotnet/coreclr/issues/25008>